sgkit_bgen/bgen_reader.py

```diff
@@ -0,0 +1,186 @@
+"""BGEN reader implementation (using bgen_reader)"""
+from pathlib import Path
+from typing import Union
+
+import dask.array as da
+import numpy as np
+from bgen_reader._bgen_file import bgen_file
+from bgen_reader._bgen_metafile import bgen_metafile
+from bgen_reader._metafile import create_metafile
+from bgen_reader._reader import _infer_metafile_filepath
+from bgen_reader._samples import generate_samples, read_samples_file
+from xarray import Dataset
+
+from sgkit import create_genotype_dosage_dataset
+from sgkit.utils import encode_array
+
+PathType = Union[str, Path]
+
+
+def _to_dict(df, dtype=None):
+    return {
+        c: df[c].to_dask_array(lengths=True).astype(dtype[c] if dtype else df[c].dtype)
+        for c in df
+    }
+
+
+VARIANT_FIELDS = [
+    ("id", str, "U"),
+    ("rsid", str, "U"),
+    ("chrom", str, "U"),
+    ("pos", str, "int32"),
+    ("nalleles", str, "int8"),
+    ("allele_ids", str, "U"),
+    ("vaddr", str, "int64"),
+]
+VARIANT_DF_DTYPE = dict([(f[0], f[1]) for f in VARIANT_FIELDS])
+VARIANT_ARRAY_DTYPE = dict([(f[0], f[2]) for f in VARIANT_FIELDS])
+
+
+class BgenReader:
+
+    name = "bgen_reader"
+
+    def __init__(self, path, persist=True, dtype=np.float32):
+        self.path = Path(path)
+
+        self.metafile_filepath = _infer_metafile_filepath(Path(self.path))
+        if not self.metafile_filepath.exists():
+            create_metafile(path, self.metafile_filepath, verbose=False)
+
+        with bgen_metafile(self.metafile_filepath) as mf:
+            self.n_variants = mf.nvariants
+            self.npartitions = mf.npartitions
+            self.partition_size = mf.partition_size
+
+            df = mf.create_variants()
+            if persist:
+                df = df.persist()
+            variant_arrs = _to_dict(df, dtype=VARIANT_ARRAY_DTYPE)
+
+            self.variant_id = variant_arrs["id"]
+            self.contig = variant_arrs["chrom"]
+            self.pos = variant_arrs["pos"]
+
+            def split_alleles(alleles, block_info=None):
+                if block_info is None or len(block_info) == 0:
+                    return alleles
+
+                def split(allele_row):
+                    return np.array(allele_row[0].split(",")[:2], dtype="S1")
```

Comment:
I think an error would be smart here if `allele_row[0]` splits into more than two alleles.  Similarly, not catching 0 or 1 alleles would probably result in a confusing error from `apply_along_axis`, so a separate check for that would probably be a big help to our future selves.

Reply:
Done in 344e9070e8617762fe938b4fa63450f5305a8ae8